File: src/model_state.cc
#include<ros/ros.h>

#include<ctime>
#include<cstdio>
#include<cstdlib>
#include<stdlib.h>
#include<pwd.h>
#include<stdio.h>
#include<iostream>
#include<sys/types.h>
#include<string>
#include<vector>
#include<math.h>

#include<std_msgs/String.h>
#include<nav_msgs/Odometry.h>
#include<geometry_msgs/Twist.h>
#include<geometry_msgs/Point.h>
#include<gazebo_msgs/ModelState.h>
#include<gazebo_msgs/ModelStates.h>
#include<gazebo_msgs/SetModelState.h>
#include<gazebo_msgs/GetModelState.h>
#include<ignition/math/Vector3.hh>
#include<ignition/math/Box.hh>

#include "mymath.cc"
using namespace std;


class SetModelPose
{
    private:
        ros::NodeHandle* rosnode; //Create ROS Node Handle
        ros::Subscriber sub; // Subscriber for subscribing to topic that produces Odometry information to be set.
        ros::Publisher pub; // This will always be /gazebo/set_model_state
        
        // Check if subscriber has been created
        bool isSubscribed;
        
        // False if no message has receive, true if any message on `sub` has
        // been received, until `pub` publishes, and then it becomes false again
        bool newMessage;
        
        // A flag to enable sending twist along with poses.
        bool enableTwist;

        bool enablePos;
        // Get the namespace (value in the group tag)
        string ns;
        
        // Get the node name
        string nodename;
        
        // Variables to store messages
        geometry_msgs::Point pose;
        geometry_msgs::Quaternion orientation;
        geometry_msgs::Twist twist;
        gazebo_msgs::ModelState modelState;

    public:
        SetModelPose()
        {
            this->rosnode = new ros::NodeHandle(""); 
            this->newMessage = false;
            this->enableTwist = false;
            this->enablePos = false;
            this->isSubscribed = false;
            
            this->ns = ros::this_node::getNamespace();
            this->ns = this->ns.substr(1, this->ns.length()-1);
            
            this->nodename = ros::this_node::getName();
            
            this->pub = this->rosnode->advertise<gazebo_msgs::ModelState>("/gazebo/set_model_state", 1);          
            // Initialize the message variable
            this->pose.x = this->pose.y = this->pose.z = 0.0;
            this->pose.z = 0.006510;
            this->orientation.x = this->orientation.y =  this->orientation.z = this->orientation.w = 0.0;
            memset(&(this->modelState), 0, sizeof(this->modelState));
            memset(&(this->modelState), 0, sizeof(this->modelState));
        }
        
        
        void createSubscriber(string odomVelTopic = "setvel")
        {
            this->sub = this->rosnode->subscribe(odomVelTopic, 1, &SetModelPose::setVelCallBack, this);
            this->isSubscribed = true;
        }
        

    public:
    
        bool subscribed()
        {
            return this->isSubscribed;
        }
        void setVelCallBack(const nav_msgs::OdometryConstPtr& _msg)
        {
            this->pose.x = _msg->pose.pose.position.x;
            this->pose.y = _msg->pose.pose.position.y;
            this->pose.z = _msg->pose.pose.position.z;
            this->orientation = _msg->pose.pose.orientation;
            this->newMessage = true;

            this->twist = _msg->twist.twist;
            

            this->modelState.model_name = this->ns;
            
            this->rosnode->param(this->nodename +"/"+"enablePos", this->enablePos, true);

            if (this->enablePos)
            {
                this->modelState.pose.position = this->pose ;//getModelState.response.pose.position;
                this->modelState.pose.orientation = this->orientation; //getModelState.response.pose.orientation;
            }
            this->rosnode->param(this->nodename +"/"+"enableTwist", this->enableTwist, false);
            
            
            if(this->enableTwist)
            {
                this->modelState.twist = this->twist;
            }
            this->modelState.reference_frame = "world";
        }
        
    public:
        void publishModelState(unsigned int rate)
        {
            ros::Rate loop_rate(rate);
            while( ros::ok() )
            {
                if( this->newMessage )
                {
                    this->pub.publish(this->modelState);
                    this->newMessage = false;
                }
                ros::spinOnce( );
                loop_rate.sleep( );
            }
        }
};


int main(int argc, char **argv)
{

    // Before you do anything, do the ros init
    ros::init(argc, argv, "SparkleSetPose");
    
    std::string enableTwist;
    std::string enablePos;
    

    //Create an object of type `SetModelPose`
    SetModelPose *sparkleModel = new SetModelPose();
    
    
    sparkleModel->createSubscriber();
    
    if (sparkleModel->subscribed())
    {
        sparkleModel->publishModelState(100);
    }
    return EXIT_SUCCESS;
    
}

File: src/model_state.py
#!/usr/bin/env python
# Author: Rahul Bhadani
# Maintainer Email: rahulbhadani@email.arizona.edu
# Initial Date: 25th August 2019
# What is this: A python script for service call and setting gazebo model positions

import rospy
import numpy as np
from std_msgs.msg import String, Header
from geometry_msgs.msg  import Twist, Pose, PoseStamped
from nav_msgs.msg import Path, Odometry
from geometry_msgs.msg import Point, Quaternion
from gazebo_msgs.msg import ModelState
import sys, getopt
import matplotlib.pylab as plt
from mpl_toolkits.mplot3d import Axes3D
from matplotlib import cm
import matplotlib.animation as animation
import time
import sys
from matplotlib.ticker import LinearLocator, FormatStrFormatter
import scipy.special as sp
from gazebo_msgs.srv import GetModelState

class SetVehicleState:

    def __init__(self, ns):
        self.ns = ns
        rospy.init_node('yawplot', anonymous=True)

        self.newMessage = False

        self.Time = rospy.Time.now()
        self.pose = []
        self.orientation = []
        self.twist = []

        self.coordinates = []
        self.sub = rospy.Subscriber('setvel'.format(ns), Odometry, self.setVel_callback)
        self.pub = rospy.Publisher('/gazebo/set_model_state', ModelState, queue_size = 1)

    def setVel_callback(self, data):
        self.newMessage = True
        self.Time = rospy.Time.now()

        self.pose = data.pose.pose.position;
        self.orientation = self.calcQuaternion(data.pose.pose.orientation.x,
                                data.pose.pose.orientation.y,
                                data.pose.pose.orientation.z)
        self.twist = data.twist.twist

    def calcQuaternion(self, roll, pitch, yaw):
        cy = np.cos(yaw * 0.5)
        sy = np.sin(yaw * 0.5)
        cp = np.cos(pitch * 0.5)
        sp = np.sin(pitch * 0.5)
        cr = np.cos(roll * 0.5)
        sr = np.sin(roll * 0.5)

        w = cy * cp * cr + sy * sp * sr
        x = cy * cp * sr - sy * sp * cr
        y = sy * cp * sr + cy * sp * cr
        z = sy * cp * cr - cy * sp * sr

        q = Quaternion(x, y, z, w)

        return q

    def setState(self):
        try:
            model_coordinates = rospy.ServiceProxy('/gazebo/get_model_state', GetModelState)
            self.coordinates = model_coordinates(self.ns[1:], 'world')

            [qX, qY, qZ, qW] = [self.coordinates.pose.orientation.x, self.coordinates.pose.orientation.y, self.coordinates.pose.orientation.z, self.coordinates.pose.orientation.w]

            [pX, pY, pZ] = [self.coordinates.pose.position.x, self.coordinates.pose.position.y, self.coordinates.pose.position.z]

            if self.newMessage is True:
               gm = ModelState()
               gm.model_name = self.ns[1:]
               gm.reference_frame = 'world'
               gm.twist = self.twist


               '''
               gm.pose.position.x = pX
               gm.pose.position.y = pY
               gm.pose.position.z = pZ
               gm.pose.orientation.x = qX
               gm.pose.orientation.y = qY
               gm.pose.orientation.z = qZ
               gm.pose.orientation.w = qW
               '''

               gm.pose.position = self.pose
               gm.pose.orientation = self.orientation

               self.pub.publish(gm)
        except rospy.ServiceException as e:
            rospy.loginfo("Get Model State service failed: {0}".format(e))



def main(argv):
    ns = rospy.get_namespace() # Retrieve namespace this way appends '/' at the end as well
    ns = ns[0:-1]
    node = SetVehicleState(ns)

    rate = rospy.Rate(60)

    rospy.loginfo("model_state.py: Namespace acquired is " + ns)

    while not rospy.is_shutdown():
        node.setState()
        rate.sleep()

if __name__ == '__main__':
    main(sys.argv[1:])


